// priority-queue-typed/dist/types/common.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FamilyPosition = exports.IterationType = exports.CP = exports.BSTVariant = void 0;
var BSTVariant;
(function (BSTVariant) {
    BSTVariant["MIN"] = "MIN";
    BSTVariant["MAX"] = "MAX";
})(BSTVariant || (exports.BSTVariant = BSTVariant = {}));
var CP;
(function (CP) {
    CP["lt"] = "lt";
    CP["eq"] = "eq";
    CP["gt"] = "gt";
})(CP || (exports.CP = CP = {}));
/**
 * Enum representing different loop types.
 *
 * - `iterative`: Indicates the iterative loop type (with loops that use iterations).
 * - `recursive`: Indicates the recursive loop type (with loops that call themselves).
 */
var IterationType;
(function (IterationType) {
    IterationType["ITERATIVE"] = "ITERATIVE";
    IterationType["RECURSIVE"] = "RECURSIVE";
})(IterationType || (exports.IterationType = IterationType = {}));
var FamilyPosition;
(function (FamilyPosition) {
    FamilyPosition["ROOT"] = "ROOT";
    FamilyPosition["LEFT"] = "LEFT";
    FamilyPosition["RIGHT"] = "RIGHT";
    FamilyPosition["ROOT_LEFT"] = "ROOT_LEFT";
    FamilyPosition["ROOT_RIGHT"] = "ROOT_RIGHT";
    FamilyPosition["ISOLATED"] = "ISOLATED";
    FamilyPosition["MAL_NODE"] = "MAL_NODE";
})(FamilyPosition || (exports.FamilyPosition = FamilyPosition = {}));
