// data-structure-typed/dist/mjs/types/common.js
export var BSTVariant;
(function (BSTVariant) {
    BSTVariant["MIN"] = "MIN";
    BSTVariant["MAX"] = "MAX";
})(BSTVariant || (BSTVariant = {}));
export var CP;
(function (CP) {
    CP["lt"] = "lt";
    CP["eq"] = "eq";
    CP["gt"] = "gt";
})(CP || (CP = {}));
/**
 * Enum representing different loop types.
 *
 * - `iterative`: Indicates the iterative loop type (with loops that use iterations).
 * - `recursive`: Indicates the recursive loop type (with loops that call themselves).
 */
export var IterationType;
(function (IterationType) {
    IterationType["ITERATIVE"] = "ITERATIVE";
    IterationType["RECURSIVE"] = "RECURSIVE";
})(IterationType || (IterationType = {}));
export var FamilyPosition;
(function (FamilyPosition) {
    FamilyPosition["ROOT"] = "ROOT";
    FamilyPosition["LEFT"] = "LEFT";
    FamilyPosition["RIGHT"] = "RIGHT";
    FamilyPosition["ROOT_LEFT"] = "ROOT_LEFT";
    FamilyPosition["ROOT_RIGHT"] = "ROOT_RIGHT";
    FamilyPosition["ISOLATED"] = "ISOLATED";
    FamilyPosition["MAL_NODE"] = "MAL_NODE";
})(FamilyPosition || (FamilyPosition = {}));
